### PR TITLE
Slot builder

### PR DESCRIPTION
**Pull request by benbierens**

### CODE REVIEW

**Review by tbekas — CHANGES_REQUESTED** [review]
Overall looks good. Please take care about error handling. We're losing too much errors in logs.

File: codex/slotbuilder/slotbuilder.nim [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+
+  for datasetBlockIndex in strategy.getIndicies(datasetSlotIndex):
+    without slotBlockCid =? await self.blockStore.getCid(datasetTreeCid, datasetBlockIndex), err:
+      error "Failed to get block CID for tree at index", index=datasetBlockIndex, tree=datasetTreeCid
+      return failure(err)
+    cids.add(slotBlockCid)
+
+  return success(cids)
+
+proc findNextPowerOfTwo*(i: int): int =
```
[the remark this answers is not in the document]
Comment (benbierens) [reply codex/slotbuilder/slotbuilder.nim:64]: How `on God's green Earth` did I not see that?! I did look for such a function... oh, well, this is what reviews are for. Thanks and fixed. :D

File: codex/slotbuilder/slotbuilder.nim [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+  return success()
+
+proc addPadBlocksToTreeBuilder(self: SlotBuilder, builder: var MerkleTreeBuilder, nBlocks: int): ?!void =
+  without cid =? emptyCid(self.manifest.version, self.manifest.hcodec, self.manifest.codec), err:
+    error "Unable to initialize empty cid"
+    return failure(err)
+
+  without emptyLeaf =? cid.mhash:
+      error "Failed to get leaf hash from empty CID"
+      return failure("Failed to get leaf hash from empty CID")
```
[the remark this answers is not in the document]
Comment (benbierens) [reply codex/slotbuilder/slotbuilder.nim:102]: Fixed indenting. Fixed lost error here, and similar lost error at line 73.

File: codex/slotbuilder/slotbuilder.nim [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+    error "Failed to add slot blocks to tree builder"
+    return failure("Failed to add slot blocks to tree builder")
+
+  if self.addPadBlocksToTreeBuilder(builder, numberOfPadBlocks).isErr:
+    error "Failed to add padding blocks to tree builder"
+    return failure("Failed to add padding blocks to tree builder")
+
+  without slotTree =? builder.build(), err:
+    error "Failed to build slot tree"
+    return failure(err)
```
[the remark this answers is not in the document]
Comment (benbierens) [reply codex/slotbuilder/slotbuilder.nim:128]: The tests agree with you. That saves us a lot of code. Excellent!

File: codex/slotbuilder/slotbuilder.nim [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+      return failure("Failed to get leaf hash from empty CID")
+
+  for i in 0 ..< nBlocks:
+    if builder.addLeaf(emptyLeaf).isErr:
+      error "Failed to add empty leaf to slot tree builder"
+      return failure("Failed to add empty leaf to slot tree builder")
+
+  return success()
+
+proc buildSlotTree*(self: SlotBuilder, slotBlocks: seq[Cid], numberOfPaddingCells: int): Future[?!MerkleTree] {.async.} =
```
[the remark this answers is not in the document]
Comment (benbierens) [reply codex/slotbuilder/slotbuilder.nim:111]: It does not. :D

File: codex/slotbuilder/slotbuilder.nim [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+proc selectSlotBlocks*(self: SlotBuilder, datasetSlotIndex: int): Future[?!seq[Cid]] {.async.} =
+  var cids = newSeq[Cid]()
+  let
+    datasetTreeCid = self.manifest.treeCid
+    blockCount = self.manifest.blocksCount
+    numberOfSlots = self.manifest.ecK
+    strategy = SteppedIndexingStrategy.new(0, blockCount - 1, numberOfSlots)
+
+  for datasetBlockIndex in strategy.getIndicies(datasetSlotIndex):
+    without slotBlockCid =? await self.blockStore.getCid(datasetTreeCid, datasetBlockIndex), err:
```
[the remark this answers is not in the document]
Comment (benbierens) [reply codex/slotbuilder/slotbuilder.nim:57]: That's a good point. I am adding `await sleepAsync(10.millis)` with a big TODO, for when threading is in place.
I really don't want to lock up the main thread, cause network traffic to get screwed, and cause all my precious distributed tests to fail.

File: codex/slotbuilder/slotbuilder.nim [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+    T: type SlotBuilder,
+    blockStore: BlockStore,
+    manifest: Manifest
+): ?!SlotBuilder =
+
+  if not manifest.protected:
+    return failure("Can only create SlotBuilder using protected manifests.")
+
+  if (manifest.blocksCount mod manifest.ecK) != 0:
+    return failure("Number of blocks must be divisable by number of slots.")
```
[the remark this answers is not in the document]
Comment (benbierens) [reply codex/slotbuilder/slotbuilder.nim:33]: At the moment, that is how it's built. It's not technically a restriction of erasure-coding. But it does keep the logic much simpler.

File: codex/slotbuilder/slotbuilder.nim (hunk 5 above)
[the remark this answers is not in the document]
Comment (tbekas) [reply codex/slotbuilder/slotbuilder.nim:58]: Good idea 👍

**Review by tbekas — APPROVED** [review]
Look good!

File: codex/slotbuilder/slotbuilder.nim [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+proc selectSlotBlocks*(self: SlotBuilder, datasetSlotIndex: int): Future[?!seq[Cid]] {.async.} =
+  var cids = newSeq[Cid]()
+  let
+    datasetTreeCid = self.manifest.treeCid
+    blockCount = self.manifest.blocksCount
+    numberOfSlots = self.manifest.ecK
+    strategy = SteppedIndexingStrategy.new(0, blockCount - 1, numberOfSlots)
+
+  for datasetBlockIndex in strategy.getIndicies(datasetSlotIndex):
+    without slotBlockCid =? await self.blockStore.getCid(datasetTreeCid, datasetBlockIndex), err:
```
Comment (tbekas) [inline codex/slotbuilder/slotbuilder.nim:58]: One more comment thought from me:
 
As far as I understand, the purpose of this change is to build a tree for slot data. Currently you're using regular merkle tree (sha256), but the idea is to plug in poseidon2 tree whenever the implementation is ready, right?

If that's the case, we won't ever use the `getCid` function, we will use `getBlock(datasetTreeCid, index)` instead. So is it actually worth changing the interface of the `Blockstore` by adding yet another method (`getCid`) that soon will be rendered unused?

File: codex/slotbuilder/slotbuilder.nim (hunk 6 above)
[the remark this answers is not in the document]
Comment (tbekas) [reply codex/slotbuilder/slotbuilder.nim:34]: Following up on yesterday conversation: number of slots = N. Where N = K + M.

File: codex/slotbuilder/slotbuilder.nim (hunk 7 above)
Comment (benbierens) [reply codex/slotbuilder/slotbuilder.nim:58]: Yeah that's a good point, and the reason why I'm not eager to merge this one yet. I want to swap out the merkle tree as this code is currently using it for the correct poseidon one first. That's still in progress. So this branch, and also the data-sampler, will have to float around for a while until that's settled.

File: codex/slotbuilder/slotbuilder.nim (hunk 6 above)
[the remark this answers is not in the document]
Comment (benbierens) [reply codex/slotbuilder/slotbuilder.nim:34]: Just to make sure it's clear everywhere, I've pulled `numberOfSlots` into the Manifest as a proc of K + M.

**Comment by benbierens** [comment]
Merged by Dmitriy in #666

**Pull request closed without merging**